src/java/com/puppetlabs/http/client/impl/JavaClient.java

```diff
@@ -493,4 +511,56 @@ public static Object coerceBodyType(InputStream body, ResponseBodyType as,
 
         return response;
     }
+
+    private static Timer.Context startTimer(MetricRegistry registry, HttpRequest request) {
+        if (registry != null) {
+            final RequestLine requestLine = request.getRequestLine();
+            final String name = MetricRegistry.name("puppetlabs.http-client", requestLine.getUri(),
+                    requestLine.getMethod());
+            return registry.timer(name).time();
+        } else {
+            return null;
+        }
+    }
+
+    private static Timer.Context startUnbufferedStreamTimer(MetricRegistry registry, HttpRequest request) {
+        if (registry != null) {
+            final RequestLine requestLine = request.getRequestLine();
+            final String name = MetricRegistry.name("puppetlabs.http-client", requestLine.getUri(),
+                    requestLine.getMethod(), "unbuffered_stream");
+            return registry.timer(name).time();
+        } else {
+            return null;
+        }
+    }
+
+    public static SortedMap<String, Timer> getClientMetrics(MetricRegistry metricRegistry){
+        if (metricRegistry != null) {
+            return metricRegistry.getTimers(new ClientMetricFilter());
+        } else {
+            return null;
+        }
+    }
+
+    public static Map<String, Map<String, Object>> getClientMetricsData(MetricRegistry metricRegistry){
+        Map<String, Map<String, Object>> metricsData = new HashMap<>();
+        SortedMap<String, Timer> timers = getClientMetrics(metricRegistry);
+        if (timers != null) {
+            for (SortedMap.Entry<String, Timer> entry : timers.entrySet()) {
+                Timer timer = entry.getValue();
+                String metricId = entry.getKey();
+                Double mean = timer.getSnapshot().getMean();
+                Long meanMillis = TimeUnit.NANOSECONDS.toMillis(mean.longValue());
+                Long count = timer.getCount();
+
+                Map<String, Object> data = new HashMap<>();
+                data.put("metric-id", metricId);
+                data.put("count", count);
+                data.put("mean", meanMillis);
+                data.put("aggregate", count * meanMillis);
+                metricsData.put(metricId, data);
```

Comment:
I'm a little waffly on this; it might be better to create a new class to hold the 4 fields and then put that class into a map instead of using `Object`.  Since we know that we're mostly going to be consuming this from Clojure I feel *slightly* less strongly about that, but I still think it might be preferable.